libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs

```diff
@@ -0,0 +1,181 @@
+failT ss = Typed (Left ss)
+
+class LiftT x where
+  liftT :: x -> Typed x
+  dropT :: Typed x -> x
+
+instance LiftT SumCond where
+  liftT (CondNever xs) = failT xs
+  liftT x = pure x
+  dropT (Typed (Left s)) = CondNever s
+  dropT (Typed (Right x)) = x
+
+instance LiftT Size where
+  liftT (SzNever xs) = failT xs
+  liftT x = pure x
+  dropT (Typed (Left s)) = SzNever s
+  dropT (Typed (Right x)) = x
+
+-- ==================================
+
+sameRep :: Rep era i -> Rep era j -> Typed (i :~: j)
+sameRep r1 r2 = case testEql r1 r2 of
+  Just x -> pure x
+  Nothing -> failT ["Type error in sameRep:\n  " ++ show r1 ++ " =/=\n  " ++ show r2]
+
+explain :: String -> Typed a -> Typed a
+explain s (Typed (Left ss)) = Typed (Left (s : ss))
+explain _ (Typed (Right x)) = Typed (Right x)
+
+mergeExplain :: (Monoid x, LiftT x) => String -> x -> x -> x
+mergeExplain message x y = dropT (explain message (liftT (x <> y)))
+
+-- ======================================================
+-- converting from a (Typed t) to something else
+
+-- The projection from (newtype Typed t = Typed {runTyped :: Either [String] x})
+-- runTyped :: Typed t => Either [String] x
+
+-- | Pushes the (Left msgs) into a call to 'error'
+errorTyped :: Typed t -> t
+errorTyped t = case runTyped t of
+  Right x -> x
+  Left xs -> error (unlines ("\nSolver-time error" : xs))
+
+-- | Pushes the (Left msgs) into a call to 'error', then injects into a Monad
+monadTyped :: Monad m => Typed t -> m t
+monadTyped t = pure (errorTyped t)
+
+-- ========================================================
+
+-- | A Pair of a (Rep era t), and a 't' . Useful to encapsulate 't' in a way
+--   one can interogate its type at runtime.
+data Dyn era where Dyn :: Rep era t -> t -> Dyn era
+
+instance Show (Dyn era) where
+  show (Dyn rep t) = "(Dyn " ++ synopsis rep t ++ ")"
+
+-- | extract the type 't' from a Dyn (if possible).
+runDyn :: Rep era t -> Dyn era -> Typed t
+runDyn rep1 (Dyn rep2 t) = do
+  Refl <- sameRep rep1 rep2
+  pure t
+
+-- | Evidence that a (Dyn era) is a (Set rng), where rng has an Ord instance.
+--   This is hard to write using (runDyn (SetR x) dyn) because SetR requires an
+--   Ord instance, and one often doesn't know any properties of 'x'
+isDynSet :: Dyn era -> Rep era rng -> Typed (HasCond Ord (Set rng))
+isDynSet (Dyn (SetR r2) set) rng = do
+  Refl <- sameRep rng r2
+  pure (With set) -- Note pattern matching against (SetR r2) provides the Ord instance
+isDynSet dyn rng = failT [show dyn ++ "is not a dynamic (Set " ++ show rng ++ ")"]
+
+-- ================================================================
+-- Computing runtime evidence of a constraint.
+
+data Id x = Id x
+
+-- | runTime evidence that the index 'i' of an indexed type '(s i)'
+--   is constrained by the class 'c'. If one has an un-indexed type
+--   't' one can always use (Id t) instead. Eg
+--   With (Id x) <- hasOrd repT (Id t)
+data HasCond c t where
+  With :: c t => (s t) -> HasCond c (s t)
```

Comment:
Would you like to make it strict?
```suggestion
  With :: c t => !(s t) -> HasCond c (s t)
```
Or remove parenthesis:
```suggestion
  With :: c t => s t -> HasCond c (s t)
```

Reply:
I don't think strictness is wanted. Often the only thing one wants is the withness of the class.